### PR TITLE
Changed DateInput to fix an issue with setting the value

### PR DESCRIPTION
**Pull request by ericsoderberghp**
#### What does this PR do?

Changed DateInput to fix an issue with setting the value

#### Where should the reviewer start?

See new unit tests to get an idea of what the changes are related to.
Then utils.js to see how the utility functions are tightened up.

#### What testing has been done on this PR?

Added unit tests
Manual interaction with DateInput Range format story

#### How should this be manually tested?

 DateInput Range format story

#### What are the relevant issues?

#5261 

#### Do the grommet docs need to be updated?

no

#### Should this PR be mentioned in the release notes?

yes

#### Is this change backwards compatible or is it a breaking change?

backwards compatible

### CODE REVIEW

File: src/js/components/DateInput/stories/RangeFormat.js [hunk 1]
```diff
@@ -26,9 +26,12 @@ export const RangeFormat = () => {
             value={value}
             format="mm/dd/yyyy-mm/dd/yyyy"
             buttonProps={{
-              label: `${dateFormat.format(
-                new Date(value[0]),
-              )} - ${dateFormat.format(new Date(value[1]))}`,
+              label:
```
Comment (taysea) [inline src/js/components/DateInput/stories/RangeFormat.js:29]: I deleted the "31" in the first date and saw this:

Screen Shot 2021-06-21 at 4 36 15 PM

Also -- are `buttonProps` doing anything in this file? The docs seems to indicate that this prop is only relevant when there is no `format`

Comment (ericsoderberghp) [reply src/js/components/DateInput/stories/RangeFormat.js]: good catches, both.
I've updated that story to not use `buttonProps`.
Also, I've fixed the stack trace issue.

File: src/js/components/DateInput/utils.js [hunk 2]
```diff
@@ -109,12 +117,13 @@ const pullDigits = (text, index) => {
...
-export const textToValue = (text, schema, valueProp) => {
-  if (!text) return undefined;
+export const textToValue = (text, schema, valueProp, range) => {
+  if (!text) return range ? [] : undefined;
 
   let result;
+
   const addDate = parts => {
-    // do a little sanity checking on the values
+    // do a little sanity checking on the values, of not ready yet, leave as is
```
Comment (taysea) [inline src/js/components/DateInput/utils.js:126]:
```suggestion
    // do a little sanity checking on the values, if not ready yet, leave as is
```

File: src/js/components/DateInput/utils.js [hunk 3]
```diff
@@ -126,20 +135,26 @@ export const textToValue = (text, schema, valueProp) => {
       parts.d > 31
     )
       return parts;
+    // parts look legit
```
Comment (taysea) [inline src/js/components/DateInput/utils.js:138]: not completely clear what this comment means

**Review by taysea — APPROVED** [review]
just a couple comments on the comments for clarity in future, but the functionality seems good.

File: src/js/components/DateInput/utils.js (hunk 3 above)
Comment (ericsoderberghp) [reply src/js/components/DateInput/utils.js]: Updated the comments in this area.

File: src/js/components/DateInput/DateInput.js [hunk 4]
```diff
@@ -102,7 +111,8 @@ const DateInput = forwardRef(
                 // clicking an edge date removes it
                 else if (range) normalizedValue = [nextValue, nextValue];
                 else normalizedValue = nextValue;
-                if (schema) setTextValue(valueToText(normalizedValue, schema));
+                if (schema)
+                  setTextValue(valueToText(normalizedValue, schema, range));
```
Comment (MikeKingdom) [inline src/js/components/DateInput/DateInput.js:115]: I'm not seeing a range arg in the utils.js valueToText. Am I missing something?

Comment (ericsoderberghp) [reply src/js/components/DateInput/DateInput.js]: good catch, fixed

**Review by ShimiSun — APPROVED** [review]

**Pull request merged**